### PR TITLE
Improve .gitignore support

### PR DESCRIPTION
**Pull request by sebastienros**
Found a bug while working on a local folder benchmark: `/obj` would be uploaded disregarding the `[Oo]bj/` rule, so implemented some tests based on the spec https://git-scm.com/docs/gitignore and fixed other issues.

### CODE REVIEW

File: src/Microsoft.Crank.Controller/Ignore/IgnoreFile.cs [hunk 1]
```diff
@@ -26,25 +20,22 @@ private IgnoreFile(string gitIgnorePath)
...
-            var ignoreFile = new IgnoreFile(path);
+            var ignoreFile = new IgnoreFile();
 
-            var currentDir = new DirectoryInfo(Path.GetDirectoryName(path));
+            var currentDir = Directory.Exists(path) ? new DirectoryInfo(path) : new DirectoryInfo(Path.GetDirectoryName(path));
             
             while (currentDir != null && currentDir.Exists)
             {
-                var gitIgnoreFilename = Path.Combine(currentDir.FullName, ".gitignore");
+                var gitIgnoreFilename = Path.GetFullPath(currentDir.FullName + "/.gitignore");
```
Comment (BrennanConroy) [inline src/Microsoft.Crank.Controller/Ignore/IgnoreFile.cs:29]: Shouldn't this still be a Path.Combine?

File: src/Microsoft.Crank.Controller/Ignore/IgnoreFile.cs [hunk 2]
```diff
@@ -26,25 +20,22 @@ private IgnoreFile(string gitIgnorePath)
...
                     var basePath = currentDir.FullName.Replace("\\", "/") + "/";
 
                     var localRules = new List<IgnoreRule>();
 
-                    // Ignore .git folders by default
-                    localRules.Add(IgnoreRule.Parse(basePath, ".git/"));
-
                     // Don't process parent folder if we are at the repository level
-                    if (Directory.Exists(Path.Combine(currentDir.FullName, ".git")))
+                    if (Directory.Exists(Path.GetFullPath(currentDir.FullName + "/.git")))
```
Comment (BrennanConroy) [inline src/Microsoft.Crank.Controller/Ignore/IgnoreFile.cs:38]: Same

File: src/Microsoft.Crank.Controller/Ignore/IgnoreFile.cs [hunk 3]
```diff
@@ -67,6 +58,28 @@ public static IgnoreFile Parse(string path, bool includeParentDirectories = fals
                                 continue;
                             }
 
+                            // Put a backslash ("\") in front of the first hash for patterns that begin with a hash.
```
Comment (BrennanConroy) [inline src/Microsoft.Crank.Controller/Ignore/IgnoreFile.cs:61]: Nit: this comment looks like it was copied from the spec, our code removes the `\` which seems odd without context.
TLDR; add/update comment
Same in IgnoreRule.cs

File: src/Microsoft.Crank.Controller/Ignore/IgnoreFile.cs (hunk 1 above)
Comment (sebastienros) [reply src/Microsoft.Crank.Controller/Ignore/IgnoreFile.cs:29]: I preferred `GetGullPath` because 
1- that's what is required, it create the fully qualified path
2- if the second argument starts with `/` it won't root it to the drive

So maybe here it's the same, but I tried to use the same pattern instead of mixing solutions

Comment (BrennanConroy) [reply src/Microsoft.Crank.Controller/Ignore/IgnoreFile.cs:29]: I meant
```
Path.GetFullPath(Path.Combine(currentDir.FullName, ".gitignore"));
```

Comment (sebastienros) [reply src/Microsoft.Crank.Controller/Ignore/IgnoreFile.cs]: Should work here. I had a unit test failing because of that but maybe it was for another place.

**Review by BrennanConroy — APPROVED** [review]
